### ISSUE TITLE
[clang-tidy] performance-unnecessary-value-param false positive

### ISSUE THREAD

**Issue by ferdinandhubbard981** [llvm/llvm-project#64930]
#### description
```cpp
#include <iostream>
#include <string>

std::string concatenateStrings(std::string baseString, std::string extraString) {
    baseString += std::move(extraString); // if extraString was a const reference, this line would involve an unnecessary copy.
    return baseString;
}
```

The reason to pass ```extraString``` by value is to potentially avoid a copy. If the caller has a ```std::string``` that they no longer need, they can ```std::move``` it into the function, and inside the function, we can move the data from it into ```baseString```

Here's an example of the above scenario:
```cpp
int main() {
    std::string str1 = "Hello, ";
    std::string str2 = "world!";

    std::string result = concatenateStrings(str1, std::move(str2));

    std::cout << result << std::endl;

    return 0;
}
```

#### clang-tidy command
```sh
clang-tidy -checks=""-*,performance-unnecessary-value-param example.cpp
```

#### expected clang-tidy output
no warnings

#### actual clang-tidy output
```
warning: the parameter 'extraString' is copied for each invocation but only used as a const reference; consider making it a const reference [performance-unnecessary-value-param]
std::string concatenateStrings(std::string baseString, std::string extraString) {
                                                                   ^
                                                       const      &
```

[full example with clang-tidy output in compiler explorer](https://godbolt.org/z/MfEhzWa69)

##### easy solution
By not outputting the warning if there is a std::move called on the parameter

##### difficult solution (not sure if possible)
checking if the std::move is redundant or not

I am working on a patch to this issue.

**Comment by firewave** [comment]
The warning is correct.
`std::string::operator+=` does not take an rvalue so it will always copy no matter if explicitly moved or not. See https://en.cppreference.com/w/cpp/string/basic_string/operator%2B%3D.

**Comment by firewave** [comment]
If you use `std::string::append()` (which shares the available overloads of `operator+=`) instead clang-tidy will also warn about the moved not happening.

```cpp
#include <string>

static std::string concatenateStrings(std::string baseString, std::string extraString) {
    baseString.append(std::move(extraString));
    return baseString;
}
```

``` 
<source>:3:75: warning: the parameter 'extraString' is copied for each invocation but only used as a const reference; consider making it a const reference [performance-unnecessary-value-param]
    3 | static std::string concatenateStrings(std::string baseString, std::string extraString) {
      |                                                                           ^
      |                                                               const      &
<source>:4:23: warning: passing result of std::move() as a const reference argument; no move will actually happen [performance-move-const-arg]
    4 |     baseString.append(std::move(extraString));
      |                       ^~~~~~~~~~
```

https://godbolt.org/z/G69eWvrod

**Comment by ferdinandhubbard981** [comment]
@firewave @EugeneZelenko 
what about the false positive in https://github.com/llvm/llvm-project/issues/62276 ?

if you change: 
```
static void set(std::string name, int value);
```
to:
```
static void set(std::string&& name, int value);
```

it goes away, but surely the warning shouldn't be there even without the &&, considering that the dissasembly shows no copy being used.

also from https://en.cppreference.com/w/cpp/string/basic_string/basic_string you can see that std::string's constructor accepts an rvalue reference:
```
basic_string( basic_string&& other ) noexcept;
```

**Comment by firewave** [comment]
In that other ticket there's just passing of the object involved but not the final usage. Depending on what `Context::set()` it might also tell you that should use a const reference.

Also things about `std::move()` are not that trivial - see #53083.
